xarray/core/coordinates.py

```diff
@@ -106,9 +107,47 @@ def to_index(self, ordered_dims: Sequence[Hashable] = None) -> pd.Index:
             (dim,) = ordered_dims
             return self._data.get_index(dim)  # type: ignore
         else:
+            from pandas.core.arrays.categorical import factorize_from_iterable
```

Comment:
Can you use `pandas.factorize` here instead? `factorize_from_iterable` is not part of pandas' public API, so  we should avoid using it.

Reply:
yes, it seems to work. This is not what is used internally by MultiIndex.from_index, but it works. It is possibly too feature-rich, because here, as we have an index, the factorisation step might be avoided. Using pd.factorize is on the safe side, probably not the performance side.

Reply:
Yes, it works. MultiIndex.from-product was not implemented with pd.factorize, but it seems to do the same job as factorize_from_iterable. Maybe it is more versatile, I didn't manage to see the difference.